backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -384,6 +390,83 @@ public FutureExperimentRunDAO(Executor executor, FutureJdbi jdbi, UAC uac) {
         .thenCompose(experimentRun -> createRoleBindingsForExperimentRun(experimentRun), executor);
   }
 
+  /**
+   * @param errorOut : Throw error while creation (true) otherwise we will keep it silent (false)
+   */
+  private ExperimentRun checkDatasetVersionBasedOnPrivileges(
+      ExperimentRun experimentRun, boolean errorOut) {
+    ExperimentRun.Builder experimentRunBuilder = experimentRun.toBuilder();
+    List<Artifact> accessibleDatasetVersions =
+        getPrivilegedDatasets(experimentRun.getDatasetsList(), errorOut);
+    experimentRunBuilder.clearDatasets().addAllDatasets(accessibleDatasetVersions);
+    return experimentRunBuilder.build();
+  }
+
+  /**
+   * @param newDatasets : new datasets for privilege check
+   * @param errorOut : Throw error while creation (true) otherwise we will keep it silent (false)
+   * @return {@link List} : accessible datasets
+   * @throws ModelDBException: modelDBException
+   */
+  private List<Artifact> getPrivilegedDatasets(List<Artifact> newDatasets, boolean errorOut) {
+    List<Artifact> accessibleDatasets = new ArrayList<>();
+    List<String> accessibleDatasetVersionIds = new ArrayList<>();
+    List<String> newDatasetLinkedIds = new ArrayList<>();
+    for (Artifact dataset : newDatasets) {
+      String datasetVersionId = dataset.getLinkedArtifactId();
+      if (!datasetVersionId.isEmpty()) {
+        newDatasetLinkedIds.add(datasetVersionId);
+      }
+    }
+
+    jdbi.useHandle(
+        handle -> {
+          Map<String, Long> datasetVersionDatasetMap = new HashMap<>();
+          var query =
+              handle
+                  .createQuery(
+                      " SELECT commit_hash, repository_id FROM repository_commit WHERE commit_hash IN (<datasetIds>) ")
+                  .bind("datasetIds", newDatasetLinkedIds);
+          query
+              .map(
+                  (rs, ctx) ->
+                      datasetVersionDatasetMap.put(
+                          rs.getString("commit_hash"), rs.getLong("repository_id")))
+              .list();
+
+          for (Artifact dataset : newDatasets) {
+            String datasetVersionId = dataset.getLinkedArtifactId();
+            if (!datasetVersionId.isEmpty()
+                && !accessibleDatasetVersionIds.contains(datasetVersionId)) {
+              try {
+                if (datasetVersionDatasetMap.containsKey(datasetVersionId)
+                    && datasetVersionDatasetMap.get(datasetVersionId) != null) {
+                  Long datasetId = datasetVersionDatasetMap.get(datasetVersionId);
+                  checkEntityPermission(
```

Comment:
You're ignoring the return of `checkEntityPermission`, which just introduces a race condition.